src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -14724,3 +14723,138 @@ void Unit::Whisper(uint32 textId, Player* target, bool isBossWhisper /*= false*/
     ChatHandler::BuildChatPacket(data, isBossWhisper ? CHAT_MSG_RAID_BOSS_WHISPER : CHAT_MSG_MONSTER_WHISPER, LANG_UNIVERSAL, this, target, bct->GetText(locale, getGender()), 0, "", locale);
     target->SendDirectMessage(&data);
 }
+
+void Unit::ValidateMovementInfo(MovementInfo* mi)
+{
+    //! Anti-cheat checks.
+    #ifdef TRINITY_DEBUG
+    #define CHECK_FOR_VIOLATING_FLAGS(check) \
+    { \
+        if (check) \
+        { \
+            TC_LOG_INFO("cheat", "Unit::ValidateMovementInfo: A violation has been detected (%s) for player GUID: %u. The player will be kicked." \
+                " Data from client: MovementFlags: %u, MovementFlags2: %u. Data in the server: MovementFlags: %u, MovementFlags2: %u.", \
+                STRINGIZE(check), GetPlayerMovingMe()->GetGUID().GetCounter(), \
+                            mi->GetMovementFlags(), mi->GetExtraMovementFlags(), \
+                            GetMovementInfo().GetMovementFlags(), GetMovementInfo().GetExtraMovementFlags()); \
+            GetPlayerMovingMe()->GetSession()->KickPlayer(); \
+            return; \
+        } \
+    }
+    #else
+    #define CHECK_FOR_VIOLATING_FLAGS(check) \
+                    if (check) \
+                    { \
+                        GetPlayerMovingMe()->GetSession()->KickPlayer(); \
+                        return; \
+                    } \
+
+    #endif 
+
+    /*! This must be a packet spoofing attempt. MOVEMENTFLAG_ROOT sent from the client is not valid
+    in conjunction with any of the moving movement flags such as MOVEMENTFLAG_FORWARD.
+    It will freeze clients that receive this player's movement info.
+    */
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_ROOT) && mi->HasMovementFlag(MOVEMENTFLAG_MASK_MOVING));
+
+    //! Cannot ascend and descend at the same time
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_ASCENDING) && mi->HasMovementFlag(MOVEMENTFLAG_DESCENDING));
+
+    //! Cannot move left and right at the same time
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_LEFT) && mi->HasMovementFlag(MOVEMENTFLAG_RIGHT));
+
+    //! Cannot strafe left and right at the same time
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_STRAFE_LEFT) && mi->HasMovementFlag(MOVEMENTFLAG_STRAFE_RIGHT));
+
+    //! Cannot pitch up and down at the same time
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_PITCH_UP) && mi->HasMovementFlag(MOVEMENTFLAG_PITCH_DOWN));
+
+    //! Cannot move forwards and backwards at the same time
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_FORWARD) && mi->HasMovementFlag(MOVEMENTFLAG_BACKWARD));
+
+    //! Cannot fly ('flying' flag) without the 'canFly' flag active
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_FLYING) && !mi->HasMovementFlag(MOVEMENTFLAG_CAN_FLY));
+
+    ////! Check for synchronisation between client and server with those server-controlled flags: 
```

Comment:
Why you do check for sync between client and server? It's really necessary?
I guess it shouldn't be here and it cause of random kicks from world.

Reply:
No the check is necessary. The goal of these checks is to ensure that the client is not adding by himself server controlled movement flags. Such as "CanFly".

The fact that it causes kicks means that I missed parts of the code that needs refactoring.